Add blog post for 4.0@rc1 (#398)
SteeltoeOSS/Documentation: docs/articles/releases/steeltoe-4-0-0-rc1.md

uid: steeltoe-4-0-0-rc1.html
tags: ["new-release"]
author.name: Tim Hess
---

# Steeltoe 4.0.0 RC1: A Major Maintenance Release

Steeltoe 4.0.0 RC1 is now available — and it's production-ready.
This release brings a modernized developer experience, compatibility with the latest .NET versions, improved performance and scalability, enhanced security and updated support for essential cloud platforms and third-party tools.
Whether you're starting something new or modernizing an existing app, 4.0 is the best way to build cloud-native .NET applications today.

## What's New Since 3.2?

Steeltoe 4.0 introduces changes across nearly every area of the framework and surrounding projects and is built for .NET 8+

For a deeper dive, check out the full changelog in the docs: [What's New in Steeltoe 4.0](https://steeltoe.io/docs/v4/welcome/whats-new.html)

### Enhanced Developer Experience

Steeltoe 4.0 brings modern .NET features and tooling improvements that make development faster and more productive:

* Streamlined packages (`*Base` and `*Core` packages have been combined), now with debug symbols embedded
* Async-first APIs
* Nullable annotations
* Embedded JSON schemas for IDE completion

> * Simplified and consistent extensions for dependency injection and configuration

Reviewer: ```suggestion
* Simplified and consistent extension methods for dependency injection and configuration
```

Almost all of the APIs Steeltoe provided were public, with little to no guidance on when and how to extend Steeltoe.
This made it harder for developers to discover the "right" approach — and it made maintenance difficult.

With Steeltoe 4.0, we've taken a deliberate approach to address this:

* **Reviewed the entire public API** to eliminate duplication, improve clarity and consistency and hide details that were not designed to be extensible.
* **Aligned APIs more closely with Microsoft conventions**, reducing surprises for all developers.
* **Added internal tooling** to detect breaking changes across releases and ensure future stability.

Breaking changes are [documented](https://steeltoe.io/docs/v4/welcome/whats-new.html), and our [samples](https://github.com/SteeltoeOSS/Samples/tree/4.x) reflect the latest best practices.

#### Steeltoe Entrypoint Changes

The majority of Steeltoe's functionality can be added to an application via `IServiceCollection` and `IConfigurationBuilder` extension methods.
In some cases, prior versions of Steeltoe also included various flavors of extension methods on host builders.

> These extensions provide the option of adding Steeltoe components with a single line of top-level code in `program.cs` and doing less work in `startup.cs`.

Reviewer: ```suggestion
These extension methods enabled adding Steeltoe components with a single line of code in `Program.cs` or `Startup.cs`.
```

> These days however, `HostApplicationBuilder` and `WebApplicationBuilder` provide simple and direct access to the configuration builder and service collection, allowing all app configuration code in `program.cs`, so the extra layer doesn't add the same reward.

Reviewer: ```suggestion
These days, however, `HostApplicationBuilder` and `WebApplicationBuilder` provide simple and direct access to the configuration builder and service collection, allowing all app configuration code in `Program.cs`, so the extra layer doesn't add the same value.
```

> For 4.0, we reviewed all extension methods to ensure we don't have situations where we don't needlessly offer multiple APIs to accomplish the same thing.
> In places where host builder extensions are definitely useful, we've made sure they're available for all currently-applicable options and worked to enhance the XML comments so it's easier for you to know which option to use.

Reviewer: ```suggestion
For 4.0, we reviewed all extension methods to ensure that we don't unnecessarily offer multiple APIs to accomplish the same thing.
In places where host builder extensions are definitely useful, we've ensured they're available for all currently applicable options and worked to enhance the XML comments so that it's easier for you to determine which option to use.
```
Suggested by Grammarly.

* Prometheus and Grafana

### Improved Documentation, Samples and Supporting Systems

Steeltoe 4.0 comes with updated docs, modernized samples, and improved tooling to help you get started faster and with more confidence. That means:

* Rewritten and reorganized samples that reflect modern .NET practices and highlight where Steeltoe fits in.
* An updated [Steeltoe Initializr](https://start.steeltoe.io/) that supports Steeltoe 4.0.
* Documentation improvements across all major areas.
* Improved test coverage and continuous integration pipelines to support faster, more reliable releases — with fewer surprises for users.
* Improved standards and automated tools for maintaining high levels of code quality.

### What's Gone

We've also used this release as an opportunity to simplify the project by removing several components that did not see great adoption and were very complex.
If you're looking for something that seems to have disappeared, check out the list of [dropped components](https://github.com/SteeltoeOSS/Steeltoe/issues/1244).

These decisions weren't taken lightly — but we believe they'll help keep Steeltoe focused and easier to use.

## Production-Ready and Supported

> Although this release is labeled a "release candidate", Steeltoe 4.0.0 RC1 reflects our intent for the final 4.0.0 release and is classified as a go-live release so that you can use it in production environments.

Reviewer: ```suggestion
Although this release is labeled a "release candidate", Steeltoe 4.0.0 RC1 reflects our intent for the final 4.0.0 release and is classified as a go-live release, allowing you to use it in production environments.
```

## Breaking Changes: Why Now?

Over the years, Steeltoe accumulated a large set of APIs that solved similar problems in slightly different ways.
Almost all of the APIs Steeltoe used were public, with little to no guidance on when and how to extend Steeltoe.
This made it harder for developers to discover the "right" approach — and it made maintenance difficult.

With Steeltoe 4.0, we've taken a deliberate approach to address this:

* **Reviewed the entire public API** to eliminate duplication, improve clarity and hide details that were not designed to be extensible
* **Aligned APIs more closely with Microsoft conventions and within Steeltoe**, reducing surprises for experienced .NET developers
* **Added tooling** to detect breaking changes across releases and ensure future stability

We know breaking changes can be frustrating, but we've tried to make this the last time it's necessary for a long while.

Every breaking change is [documented](https://steeltoe.io/docs/v4/welcome/breaking-changes.html), and our [samples](https://github.com/SteeltoeOSS/Samples/tree/4.x) reflect the latest best practices.

## Modern .NET Support: Simpler App Setup

> Steeltoe remains easy to use and integrates naturally with modern .NET patterns, now including first-class support for `WebApplicationBuilder` and `Minimal APIs`.

Author: Yeah... this probably should have been `HostApplicationBuilder`

That means you can write something like this:

> ```csharp

Author: Yeah, I'm not sure about this whole section, we might be better off without it
